distributed/deploy/tests/test_spec_cluster.py

```diff
@@ -156,6 +156,26 @@ class MyCluster(SpecCluster):
             assert len(cluster.workers) == 2
 
 
+@pytest.mark.slow
+@pytest.mark.asyncio
+async def test_restart(cleanup):
+    # Regression test for https://github.com/dask/distributed/issues/3062
+    worker = {"cls": Worker, "options": {"nthreads": 1}}
+    with dask.config.set({"distributed.deploy.lost-worker-timeout": "2s"}):
+        async with SpecCluster(
+            asynchronous=True, scheduler=scheduler, worker=worker
+        ) as cluster:
+            async with Client(cluster, asynchronous=True) as client:
+                cluster.scale(2)
+                await cluster
+                assert len(cluster.workers) == 2
+
+                await client.restart()
+                await asyncio.sleep(3)
```

Comment:
Do we need this sleep here?  I've found that fixed sleeps are bad in our tests for two reasons:

1.  They have to be *very* conservative on travis-ci, otherwise we get intermittent failures
2.  If they're very conservative then they take a long time

Can you say more about the reason for this sleep?  Perhaps there is some other thing that we could wait on or poll to make this more precise.

Reply:
Thanks for marking it slow.  I suppose that if we need the sleep then that is a nice compromise.

Reply:
The main thin we want to check here is that, after workers have been restarted, the delayed call  to `f` doesn't remove the restarted workers. My thinking was to set `distributed.deploy.lost-worker-timeout` sufficiently large such that the delayed call to `f` happens after workers are restarted. The sleep was added so we know the `f` call has already happened. 

Now that I've typed my reasoning out, it does seem a little fragile... I'll think a little more about if there's something else more direct we could wait on